---
title: ""
output: html_document
date: '2022-04-19'
---

```{r setup, include=FALSE}

library(dplyr)
library(ggplot2)
library(patchwork)
#library(tibble)
#library(vroom)
`%>%` <- magrittr::`%>%`

library(circlize)


knitr::opts_chunk$set(echo = TRUE)
```

## CNV Plots

Description here

### Data

```{r}
### read in histology file
# histology file has the metadata for each biospecimen like patient id, tumor type,
# sequenceing type, specimen type, cancer type, etc.
vroom::vroom('../OpenPedCan-analysis/data/histologies.tsv') -> histologies


### find example genes for plot drafts; one time only
# find gene with lots of deletions for examples
# histologies %>%
#   filter(cancer_group == 'Neuroblastoma') %>%
#   left_join(cnv, by = c('Kids_First_Biospecimen_ID' = 'biospecimen_id')) %>%
#   filter(!is.na(copy_number)) %>%
#   count(gene_symbol, status) %>% 
#   tidyr::pivot_wider(names_from = status, values_from = n) %>%
#   arrange(desc(`deep deletion`))
# 
# # find average gene for examples
# histologies %>%
#   filter(cancer_group == 'Neuroblastoma') %>%
#   left_join(cnv, by = c('Kids_First_Biospecimen_ID' = 'biospecimen_id')) %>%
#   filter(!is.na(copy_number)) %>%
#   count(gene_symbol, status) %>%
#   tidyr::pivot_wider(names_from = status, values_from = n) %>%
#   filter(neutral > gain, neutral > loss, neutral > `deep deletion`, neutral > amplification)

### make table of example genes for filtering
tibble::tibble(gene_symbol = c('ADNP2', 'DUX4', 'IRS4', 'MYCN'),
               ensembl = c('ENSG00000101544', 'ENSG00000260596',
                           'ENSG00000260548', 'ENSG00000134323')) -> gene_ensembl
```

```{r}
### current unused but will or may use in future (will need to include efo 
### code with disease name)
# vroom::vroom('../OpenPedCan-analysis/data/efo-mondo-map.tsv') -> efo_mondo_map
vroom::vroom('../OpenPedCan-analysis/data/ensg-hugo-pmtl-mapping.tsv') -> pmtl_genes
# vroom::vroom('../OpenPedCan-analysis/data/fusion-putative-oncogenic.tsv')

# independent specimens list; using the same one as the SNV frequencies module
# https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/dev/analyses/snv-frequencies
# as Jo Lynne requested. Some samples are duplicated and this is a list of unique
# samples that all analyses use
vroom::vroom('../OpenPedCan-analysis/data/independent-specimens.wgswxspanel.primary.prefer.wxs.tsv') -> independent_specs
vroom::vroom('../OpenPedCan-analysis/data/independent-specimens.rnaseq.primary.tsv') -> rnaseq_independent_spec

### read in CNV table
# old version of table without all neutral copy numbers
# vroom::vroom('../OpenPedCan-analysis/data/v10/consensus_wgs_plus_cnvkit_wxs.tsv') -> cnv
# updated version with CNV calls for all genes USE THIS ONE
vroom::vroom('consensus_wgs_plus_cnvkit_wxs.tsv') %>%
  # filter for masterlist of biospecimen ids, for primary tumor only
  filter(ensembl %in% gene_ensembl$ensembl,
         biospecimen_id %in% independent_specs$Kids_First_Biospecimen_ID) %>%
  left_join(histologies, by = c('biospecimen_id' = 'Kids_First_Biospecimen_ID')) %>%
  filter(sample_type == 'Tumor', broad_tumor_descriptor == 'Diagnosis',
         !is.na(cancer_group), biospecimen_id != 'TARGET-10-PARDSP-09A-01D') -> cnv

readr::read_rds('../OpenPedCan-analysis/data/gene-expression-rsem-tpm-collapsed.rds') %>%
  tibble::rownames_to_column('gene') %>% #select(gene) %>% sample_n(100) -> random_genes
  select(!contains('GTEX')) %>%
  filter(gene %in% gene_ensembl$gene_symbol) %>%
  tidyr::pivot_longer(2:ncol(.), names_to = 'biospecimen_id', values_to = 'tpm') %>% 
  inner_join(rnaseq_independent_spec, 
            by = c('biospecimen_id' = 'Kids_First_Biospecimen_ID')) %>%
  left_join(histologies, 
            by = c('biospecimen_id' = 'Kids_First_Biospecimen_ID', 
                   'Kids_First_Participant_ID')) %>%
  filter(Kids_First_Participant_ID %in% cnv$Kids_First_Participant_ID,
         sample_type == 'Tumor', broad_tumor_descriptor == 'Diagnosis',
         !is.na(cancer_group)) -> expression
```

#### Outlier

```{r}
cnv %>%
  filter(biospecimen_id == 'TARGET-10-PARDSP-09A-01D') %>%
ggplot(aes(x = copy_number)) +
  geom_histogram(bins = 20, color = 'black', fill = 'white') +
  labs(x = 'Copy Number', y = 'Number of Genes', 
       title = 'TARGET-10-PARDSP-09A-01D') +
  theme_bw(base_size = 20)

# histologies %>% filter(Kids_First_Biospecimen_ID == 'TARGET-10-PARDSP-09A-01D')
```

<br>

### Gene Page Plot

#### Option 1: Beeswarm

```{r, fig.width = 12, fig.height = 4}
for (i in gene_ensembl$gene_symbol) {
cnv %>%
  filter(gene_symbol == i) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
  mutate(copy_number2 = factor(ifelse(copy_number >= 5, '5+', copy_number),
                               levels = c('5+', '4', '3', '2', '1', '0'))) %>%
ggplot(aes(x = cancer_group, y = copy_number)) +
  ggbeeswarm::geom_quasirandom(aes(color = copy_number2), size = 1, groupOnX = T) +
    viridis::scale_color_viridis(discrete = T, option = 'viridis', direction = -1, drop = F) +
  geom_hline(yintercept = 2, linetype = 'dashed', color = 'gray60') +
  labs(x = 'Disease', y = 'Copy Number', color = 'Copy\nNumber',
       title = paste0(filter(gene_ensembl, gene_symbol == i)$gene_symbol, ' (',
                      filter(gene_ensembl, gene_symbol == i)$ensembl, ')')) +
  theme_classic(base_size = 20) + 
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1, size = 6),
        plot.margin = margin(10, 10, 10, 50)) -> p1
  print(p1)
  # ggsave(paste0('gene_plots/beeswarm_', i, '.png'), p1)
  # ggsave(paste0('gene_plots/beeswarm_UPDATED_', i, '.png'), p1)
}
```

#### Option 2: Boxplots

```{r, fig.width = 12, fig.height = 4}
for (i in gene_ensembl$gene_symbol) {
cnv %>%
  filter(gene_symbol == i) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
ggplot(aes(x = cancer_group, y = copy_number)) +
  geom_boxplot() +
  geom_hline(yintercept = 2, linetype = 'dashed', color = 'gray60') +
  labs(x = 'Disease', y = 'Copy Number', 
       title = paste0(filter(gene_ensembl, gene_symbol == i)$gene_symbol, ' (',
                      filter(gene_ensembl, gene_symbol == i)$ensembl, ')')) +
  theme_classic(base_size = 20) + 
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1, size = 8),
        plot.margin = margin(10, 10, 10, 50)) -> p2
  print(p2)
  # ggsave(paste0('gene_plots/boxplot_', i, '.png'), p2)
  # ggsave(paste0('gene_plots/boxplot_UPDATED_', i, '.png'), p2)
  # print(paste0('gene_plots/boxplot_', i, '.png'))
}
```

#### Option 3: Boxplots + beeswarms

```{r, fig.width = 12, fig.height = 4}
for (i in gene_ensembl$gene_symbol) {
cnv %>%
  filter(gene_symbol == i) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
  mutate(copy_number2 = factor(ifelse(copy_number >= 5, '5+', copy_number),
                               levels = c('5+', '4', '3', '2', '1', '0'))) %>%
ggplot(aes(x = cancer_group, y = copy_number)) +
  geom_boxplot(alpha = 0) +
  ggbeeswarm::geom_quasirandom(aes(color = copy_number2), size = 1, groupOnX = T) +
  viridis::scale_color_viridis(discrete = T, option = 'viridis', direction = -1, drop = F) +
  geom_hline(yintercept = 2, linetype = 'dashed', color = 'gray60') +
  labs(x = 'Disease', y = 'Copy Number', color = 'Copy\nNumber',
       title = paste0(filter(gene_ensembl, gene_symbol == i)$gene_symbol, ' (',
                      filter(gene_ensembl, gene_symbol == i)$ensembl, ')')) +
  theme_classic(base_size = 20) + 
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1, size = 6),
        plot.margin = margin(10, 10, 10, 50)) -> p3a
  print(p3a)
  # ggsave(paste0('gene_plots/boxplot_beeswarm_', i, '.png'), p3)
  # ggsave(paste0('gene_plots/boxplot_beeswarm_UPDATED_', i, '.png'), p3a)
  # print(paste0('gene_plots/boxplot_beeswarm_', i, '.png'))
}
```

Or variation 2, violin plots + beeswarms

```{r, fig.width = 12, fig.height = 4}
for (i in gene_ensembl$gene_symbol) {
cnv %>%
  filter(gene_symbol == i) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
  mutate(copy_number2 = factor(ifelse(copy_number >= 5, '5+', copy_number),
                               levels = c('5+', '4', '3', '2', '1', '0'))) %>%
ggplot(aes(x = cancer_group, y = copy_number)) +
  geom_violin(fill = 'gray80') +
  ggbeeswarm::geom_quasirandom(aes(color = copy_number2), size = 1, groupOnX = T) +
  viridis::scale_color_viridis(discrete = T, option = 'viridis', direction = -1, drop = F) +
  geom_hline(yintercept = 2, linetype = 'dashed', color = 'gray60') +
  labs(x = 'Disease', y = 'Copy Number', color = 'Copy\nNumber',
       title = paste0(filter(gene_ensembl, gene_symbol == i)$gene_symbol, ' (',
                      filter(gene_ensembl, gene_symbol == i)$ensembl, ')')) +
  theme_classic(base_size = 20) + 
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1, size = 6),
        plot.margin = margin(10, 10, 10, 50)) -> p3b
  print(p3b)
  # ggsave(paste0('gene_plots/violin_beeswarm_', i, '.png'), p3b)
  # print(paste0('gene_plots/violin_beeswarm_', i, '.png'))
}
```


#### Option 4: Stacked bar plots

1. Raw sample counts with number label (add note in powerpoint that this could also be plotted with qualitative label)
2. Log10 sample counts
3. Percentage summing to 100

---

Raw sample counts

```{r, fig.width = 12, fig.height = 4}
for (i in gene_ensembl$gene_symbol) {
cnv %>%
  filter(gene_symbol == i) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
  mutate(copy_number2 = factor(ifelse(copy_number >= 5, '5+', copy_number),
                               levels = c('5+', '4', '3', '2', '1', '0'))) %>%
ggplot(aes(x = cancer_group, fill = copy_number2)) +
  geom_bar() +
  viridis::scale_fill_viridis(discrete = T, option = 'viridis', direction = -1, drop = F) +
  labs(x = 'Disease', y = 'Number of Samples', fill = 'Copy Number',
       title = paste0(filter(gene_ensembl, gene_symbol == i)$gene_symbol, ' (',
                      filter(gene_ensembl, gene_symbol == i)$ensembl, ')')) +
  theme_classic(base_size = 12) + 
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1, size = 8),
        plot.margin = margin(10, 10, 10, 50)) -> p4a
  print(p4a)
  # ggsave(paste0('gene_plots/bar_plot_counts_', i, '.png'), p4a)
  # ggsave(paste0('gene_plots/bar_plot_counts_UPDATED_', i, '.png'), p4a)
  # print(paste0('gene_plots/bar_plot_counts_', i, '.png'))
}
```

Log10 sample counts

```{r, fig.width = 12, fig.height = 4}
for (i in gene_ensembl$gene_symbol) {
cnv %>%
  filter(gene_symbol == i) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
  mutate(copy_number2 = factor(ifelse(copy_number >= 5, '5+', copy_number),
                               levels = c('5+', '4', '3', '2', '1', '0'))) %>%
ggplot(aes(x = cancer_group, fill = copy_number2)) +
  geom_bar() +
  scale_y_log10() +
  viridis::scale_fill_viridis(discrete = T, option = 'viridis', direction = -1, drop = F) +
  labs(x = 'Disease', y = 'Log10 Number of Samples', fill = 'Copy Number',
       title = paste0(filter(gene_ensembl, gene_symbol == i)$gene_symbol, ' (',
                      filter(gene_ensembl, gene_symbol == i)$ensembl, ')')) +
  theme_classic(base_size = 12) + 
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1, size = 8),
        plot.margin = margin(10, 10, 10, 50)) -> p4b
  print(p4b)
  # ggsave(paste0('gene_plots/bar_plot_log10counts_', i, '.png'), p4b)
  # ggsave(paste0('gene_plots/bar_plot_log10counts_UPDATED_', i, '.png'), p4b)
  # print(paste0('gene_plots/bar_plot_log10counts_', i, '.png'))
}
```

Percentages

```{r, fig.width = 12, fig.height = 4}
for (i in gene_ensembl$gene_symbol) {
cnv %>%
  filter(gene_symbol == i) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
  mutate(copy_number2 = factor(ifelse(copy_number >= 5, '5+', copy_number),
                               levels = c('5+', '4', '3', '2', '1', '0'))) %>%
  count(cancer_group, copy_number2, n) %>% 
  mutate(percent = round((nn / n) * 100),
         label = paste0(percent, '%'),
         label = ifelse(label == '0%', '<1%', label)) %>%
  group_by(cancer_group) %>% 
  mutate(rownum = row_number()) %>%
  ungroup() %>%
  mutate(count_label = ifelse(rownum == 1, n, '')) %>%
ggplot(aes(x = cancer_group, y = percent, fill = copy_number2)) +
  geom_col() +
  viridis::scale_fill_viridis(discrete = T, option = 'viridis', direction = -1, drop = F) +
  geom_text(aes(label = label, color = copy_number2), 
            position = position_stack(vjust = 0.5), size = 2) +
  scale_color_manual(values = c('gray0', 'gray20', 'gray40', 'gray60', 
                                'gray80', 'gray100'),
                     guide = 'none', drop = F) +
  geom_text(aes(label = count_label), y = 107) +
  coord_cartesian(ylim = c(0, 107)) +
  scale_y_continuous(breaks = c(0, 20, 40, 60, 80, 100)) +
  labs(x = 'Disease', y = 'Percentage of Samples', fill = 'Copy Number',
       title = paste0(filter(gene_ensembl, gene_symbol == i)$gene_symbol, ' (',
                      filter(gene_ensembl, gene_symbol == i)$ensembl, ')')) +
  theme_classic(base_size = 16) + 
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1, size = 8),
        plot.margin = margin(10, 10, 10, 50)) -> p4c
  print(p4c)
  # ggsave(paste0('gene_plots/bar_plot_percents_', i, '.png'), p4c)
  # ggsave(paste0('gene_plots/bar_plot_percents_UPDATED_', i, '.png'), p4c)
  # print(paste0('gene_plots/bar_plot_percents_', i, '.png'))
}
```

#### Option 5: Grid plot

Counts

```{r, fig.width = 12, fig.height = 4}
for (i in gene_ensembl$gene_symbol) {
cnv %>% 
  filter(gene_symbol == i) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
  mutate(copy_number2 = factor(ifelse(copy_number >= 5, '5+', copy_number),
                               levels = c('5+', '4', '3', '2', '1', '0'))) %>%
  count(cancer_group, copy_number2) %>%
ggplot(aes(x = cancer_group, y = copy_number2, fill = copy_number2)) +
  geom_tile(color = "white", lwd = 1.5, linetype = 1) +
  scale_y_discrete(limits = rev, drop = F) +
  coord_fixed() +
  geom_text(aes(label = n), size = 3, color = 'gray60') +
  viridis::scale_fill_viridis(discrete = T, option = 'viridis', direction = -1, drop = F) +
  scale_color_manual(values = c('gray0', 'gray20', 'gray40', 'gray60', 
                                'gray80', 'gray100'),
                     guide = 'none', drop = F) +
  labs(x = 'Disease', y = 'Copy Number', fill = 'Copy Number', 
       title = paste0(filter(gene_ensembl, gene_symbol == i)$gene_symbol, ' (',
                      filter(gene_ensembl, gene_symbol == i)$ensembl, ')')) +
  theme_bw(base_size = 12) + ### any style guidelines for plots?
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1, size = 8),
        plot.margin = margin(10, 10, 10, 50),
        panel.grid.minor = element_line(color = 'gray60')) -> p5a
  print(p5a)
  # ggsave(paste0('gene_plots/tile_plot_counts_', i, '.png'), p5a)
  # ggsave(paste0('gene_plots/tile_plot_counts_UPDATED_', i, '.png'), p5a)
  # print(paste0('gene_plots/tile_plot_counts_', i, '.png'))
}
```

Percentages

```{r, fig.width = 12, fig.height = 4}
for (i in gene_ensembl$gene_symbol) {
cnv %>% 
  filter(gene_symbol == i) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
  mutate(copy_number2 = factor(ifelse(copy_number >= 5, '5+', copy_number),
                               levels = c('5+', '4', '3', '2', '1', '0'))) %>%
  count(cancer_group, n, copy_number2) %>%
  mutate(percent = round((nn / n) * 100),
         label = paste0(percent, '%'),
         label = ifelse(label == '0%', '<1%', label)) %>%
  group_by(cancer_group) %>% 
  mutate(rownum = row_number()) %>%
  ungroup() %>%
  mutate(count_label = ifelse(rownum == 1, n, '')) %>%
ggplot(aes(x = cancer_group, y = copy_number2, fill = copy_number2)) +
  geom_tile(color = "white", lwd = 1.5, linetype = 1) +
  scale_y_discrete(limits = rev, drop = F) +
  coord_fixed(clip = 'off') +
  geom_text(aes(label = label, color = copy_number2), size = 2.5, color = 'gray60') +
  viridis::scale_fill_viridis(discrete = T, option = 'viridis', direction = -1, drop = F) +
  scale_color_manual(values = c('gray0', 'gray20', 'gray40', 'gray60', 
                                'gray80', 'gray100'),
                     guide = 'none', drop = F) +
  geom_text(aes(label = count_label), y = 7, size = 3.5) +
  labs(x = 'Disease', y = 'Copy Number', fill = 'Copy Number', 
       title = paste0(filter(gene_ensembl, gene_symbol == i)$gene_symbol, ' (',
                      filter(gene_ensembl, gene_symbol == i)$ensembl, ')\n')) +
  theme_bw(base_size = 12) + 
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1, size = 8),
        plot.margin = margin(10, 10, 10, 60),
        panel.grid.minor = element_line(color = 'gray60')) -> p5b
  print(p5b)
  # ggsave(paste0('gene_plots/tile_plot_percentages_', i, '.png'), p5b)
  # ggsave(paste0('gene_plots/tile_plot_percentages_UPDATED_', i, '.png'), p5b)
  # print(paste0('gene_plots/tile_plot_percentages_', i, '.png'))
}
```

Bubble plot size by count

````{r, fig.width = 12, fig.height = 4}
for (i in gene_ensembl$gene_symbol) {
cnv %>% 
  filter(gene_symbol == i) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
  mutate(copy_number2 = factor(ifelse(copy_number >= 5, '5+', copy_number),
                               levels = c('5+', '4', '3', '2', '1', '0'))) %>%
  count(cancer_group, copy_number2) %>%
  group_by(cancer_group) %>% 
  mutate(rownum = row_number(),
         group_count = sum(n)) %>%
  ungroup() %>%
  mutate(count_label = ifelse(rownum == 1, group_count, '')) %>%
ggplot(aes(x = cancer_group, y = copy_number2)) +
  scale_y_discrete(limits = rev, drop = F) +
  geom_point(aes(size = n, fill = copy_number2), shape = 21) +
  scale_size_continuous(guide = 'none', range = c(1, 10)) +
  geom_text(aes(label = n), size = 3, color = 'gray60') +
  viridis::scale_fill_viridis(discrete = T, option = 'viridis', direction = -1, drop = F) +
  scale_color_manual(values = c('gray0', 'gray20', 'gray40', 'gray60', 
                                'gray80', 'gray100'),
                     guide = 'none', drop = F) +
  coord_cartesian(clip = 'off') +
  geom_text(aes(label = count_label), y = 7, size = 3.5) +
  labs(x = 'Disease', y = 'Copy Number', fill = 'Copy Number', 
       title = paste0(filter(gene_ensembl, gene_symbol == i)$gene_symbol, ' (',
                      filter(gene_ensembl, gene_symbol == i)$ensembl, ')\n')) +
  theme_bw(base_size = 12) + 
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1, size = 8),
        plot.margin = margin(10, 10, 10, 50),
        panel.grid.minor = element_line(color = 'gray60')) -> p5c
  print(p5c)
  # ggsave(paste0('gene_plots/bubble_plot_counts_', i, '.png'), p5c)
}

```

Bubble plot size by percentage

````{r, fig.width = 12, fig.height = 4}
for (i in gene_ensembl$gene_symbol) {
cnv %>% 
  filter(gene_symbol == i) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
  mutate(copy_number2 = factor(ifelse(copy_number >= 5, '5+', copy_number),
                               levels = c('5+', '4', '3', '2', '1', '0'))) %>%
  count(cancer_group, n, copy_number2) %>%
  mutate(percent = round((nn / n) * 100),
         label = paste0(percent, '%'),
         label = ifelse(label == '0%', '<1%', label)) %>%
  group_by(cancer_group) %>% 
  mutate(rownum = row_number()) %>%
  ungroup() %>%
  mutate(count_label = ifelse(rownum == 1, n, '')) %>%
ggplot(aes(x = cancer_group, y = copy_number2)) +
  scale_y_discrete(limits = rev, drop = F) +
  geom_point(aes(size = percent, fill = copy_number2), shape = 21) +
  scale_size_continuous(guide = 'none', range = c(1, 10)) +
  geom_text(aes(label = label), size = 3, color = 'gray60') +
  viridis::scale_fill_viridis(discrete = T, option = 'viridis', direction = -1, drop = F) +
  scale_color_manual(values = c('gray0', 'gray20', 'gray40', 'gray60', 
                                'gray80', 'gray100'),
                     guide = 'none', drop = F) +
  coord_cartesian(clip = 'off') +
  geom_text(aes(label = count_label), y = 7, size = 3.5) +
  labs(x = 'Disease', y = 'Copy Number', fill = 'Copy Number', 
       title = paste0(filter(gene_ensembl, gene_symbol == i)$gene_symbol, ' (',
                      filter(gene_ensembl, gene_symbol == i)$ensembl, ')\n')) +
  theme_bw(base_size = 12) + 
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1, size = 8),
        plot.margin = margin(10, 10, 10, 50),
        panel.grid.minor = element_line(color = 'gray60')) -> p5d
  print(p5d)
  # ggsave(paste0('gene_plots/bubble_plot_percents_', i, '.png'), p5d)
}
```

Grid with boxplots or jitter plots on top

```{r, fig.width = 12, fig.height = 6}
for (i in gene_ensembl$gene_symbol) {
cnv %>% 
  filter(gene_symbol == i) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
  mutate(copy_number2 = factor(ifelse(copy_number >= 5, '5+', copy_number),
                               levels = c('5+', '4', '3', '2', '1', '0'))) %>%
  count(cancer_group, copy_number2) %>%
ggplot(aes(x = cancer_group, y = copy_number2, fill = copy_number2)) +
  geom_tile(color = "white", lwd = 1.5, linetype = 1) +
  scale_y_discrete(limits = rev, drop = F) +
  coord_fixed() +
  geom_text(aes(label = n), size = 3, color = 'gray60') +
  viridis::scale_fill_viridis(discrete = T, option = 'viridis', direction = -1, drop = F) +
  scale_color_manual(values = c('gray0', 'gray20', 'gray40', 'gray60', 
                                'gray80', 'gray100'),
                     guide = 'none', drop = F) +
  labs(x = 'Disease', y = 'Copy Number', fill = 'Copy Number') +
  theme_bw() + ### any style guidelines for plots?
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1, size = 8),
        plot.margin = margin(0, 10, 10, 50),
        panel.grid.minor = element_line(color = 'gray60')) -> p5e1

cnv %>%
  filter(gene_symbol == i) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
  mutate(copy_number2 = factor(ifelse(copy_number >= 5, '5+', copy_number),
                               levels = c('5+', '4', '3', '2', '1', '0'))) %>%
ggplot(aes(x = cancer_group, y = copy_number)) +
  geom_violin(fill = 'gray80') +
  ggbeeswarm::geom_quasirandom(aes(color = copy_number2), size = 1, groupOnX = T) +
  viridis::scale_color_viridis(discrete = T, option = 'viridis', direction = -1, 
                               drop = F) +
  geom_hline(yintercept = 2, linetype = 'dashed', color = 'gray60') +
  labs(y = 'Copy Number') +
  theme_classic(base_size = 20) + 
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(),
        axis.line.x = element_blank(), axis.title.x = element_blank(),
        plot.margin = margin(10, 10, 0, 50),
        legend.position = 'none') -> p5e2


(p5e2 / p5e1) + plot_layout(guides = 'collect', widths = 1) +
  plot_annotation(title = paste0('Neuroblastoma', ": ",
                  filter(gene_ensembl, gene_symbol == i)$gene_symbol, ' (',
                  filter(gene_ensembl, gene_symbol == i)$ensembl, ')')) -> p5e
print(p5e)
  ggsave(paste0('gene_plots/grid_beeswarm_UPDATED_', i, '.png'), p5e)
}
```


<br><br>

### Evidence Page Plot

Stacked bar plot

```{r}
for (i in gene_ensembl$gene_symbol) {
cnv %>%
  filter(gene_symbol == i, cancer_group == 'Neuroblastoma') %>%
  mutate(status = ifelse(!is.na(status), status,
                         case_when(copy_number == 0 ~ 'deep deletion',
                         copy_number < ploidy & copy_number != 0 ~ 'loss',
                         copy_number == ploidy ~ 'neutral',
                         copy_number <= (2*ploidy) ~ 'gain',
                         copy_number > (2*ploidy) ~ 'amplification'))) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
  select(Kids_First_Participant_ID, biospecimen_id, status, copy_number, ploidy, ensembl, gene_symbol, 
         cohort) -> gene_disease

rbind(gene_disease, mutate(gene_disease, cohort = 'All Cohorts')) -> gene_disease2
  
gene_disease2 %>%
  add_count(cohort, name = 'sample_count') %>%
  count(cohort, status, sample_count) %>%
  mutate(percent = round((n / sample_count) * 100),
         label = paste0(percent, '%'),
         label = ifelse(label == '0%', '<1%', label),
         status = factor(status, levels = c('amplification', 'gain', 'neutral', 
                                            'loss', 'deep deletion'))) %>%
  group_by(cohort) %>%
  mutate(count_label = ifelse(row_number() == 1, paste0('n = ', sample_count), '')) %>%
  ungroup() %>%
ggplot(aes(x = cohort, y = percent, fill = status)) +
  geom_col() +
  viridis::scale_fill_viridis(discrete = T, option = 'viridis', direction = -1, drop = F) +
  geom_text(aes(label = label, color = status), 
            position = position_stack(vjust = 0.5)) +
  scale_color_manual(values = c('gray0', 'gray20', 'gray40', 'gray60', 
                                'gray80', 'gray100'),
                     guide = 'none', drop = F) +
  geom_text(aes(label = count_label), y = 105) +
  coord_cartesian(ylim = c(0, 107)) +
  scale_y_continuous(breaks = c(0, 20, 40, 60, 80, 100)) +
  labs(x = 'Cohort', y = 'Percent Copy Number', fill = 'Copy Number',
       title = paste0(ifelse(filter(gene_ensembl, gene_symbol == i)$gene_symbol == 'DUX4', 
                             'Diffuse midline glioma', 'Neuroblastoma'), ": ",
                      filter(gene_ensembl, gene_symbol == i)$gene_symbol, ' (',
                      filter(gene_ensembl, gene_symbol == i)$ensembl, ')')) + 
  theme_classic(base_size = 16) -> ep1
  print(ep1)
  # ggsave(paste0('gene_plots/evidence_plot_bar_percentages_', i, '.png'), ep1)
  # ggsave(paste0('gene_plots/evidence_plot_bar_percentages_UPDATED_', i, '.png'), ep1)
  # print(paste0('gene_plots/evidence_plot_bar_percentages_', i, '.png'))
}
```

Percentage bar plot + gene expression TPMs

```{r, fig.width = 10, fig.height = 4}
for (i in gene_ensembl$gene_symbol) {
### stacked percentage boxplot
  cnv %>%
  filter(gene_symbol == i, cancer_group == 'Neuroblastoma') %>%
  mutate(status = ifelse(!is.na(status), status,
                         case_when(copy_number == 0 ~ 'deep deletion',
                         copy_number < ploidy & copy_number != 0 ~ 'loss',
                         copy_number == ploidy ~ 'neutral',
                         copy_number <= (2*ploidy) ~ 'gain',
                         copy_number > (2*ploidy) ~ 'amplification'))) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
  select(Kids_First_Participant_ID, biospecimen_id, status, copy_number, ploidy, ensembl, gene_symbol, 
         cohort) -> gene_disease

rbind(gene_disease, mutate(gene_disease, cohort = 'All Cohorts')) -> gene_disease2
  
gene_disease2 %>%
  add_count(cohort, name = 'sample_count') %>%
  count(cohort, status, sample_count) %>%
  mutate(percent = round((n / sample_count) * 100),
         label = paste0(percent, '%'),
         status = factor(status, levels = c('amplification', 'gain', 'neutral', 
                                            'loss', 'deep deletion'))) %>%
  group_by(cohort) %>%
  mutate(count_label = ifelse(row_number() == 1, paste0('n = ', sample_count), '')) %>%
  ungroup() %>%
ggplot(aes(x = cohort, y = percent, fill = status)) +
  geom_col() +
  viridis::scale_fill_viridis(discrete = T, option = 'viridis', direction = -1, drop = F) +
  geom_text(aes(label = label, color = status), 
            position = position_stack(vjust = 0.5)) +
  scale_color_manual(values = c('gray0', 'gray20', 'gray40', 'gray60', 
                                'gray80', 'gray100'),
                     guide = 'none', drop = F) +
  geom_text(aes(label = count_label), y = 105) +
  coord_cartesian(ylim = c(0, 107)) +
  scale_y_continuous(breaks = c(0, 20, 40, 60, 80, 100)) +
  labs(x = 'Cohort', y = 'Percent Copy Number', fill = 'Copy Number') + 
  theme_classic(base_size = 16) -> ep2a
  # print(ep2a)
  
### gene expression boxplots
expression %>%
  filter(gene == i, cancer_group == 'Neuroblastoma') %>%
  add_count(cohort, cancer_group) %>% 
  filter(n >= 3) %>%
  select(Kids_First_Participant_ID, biospecimen_id, gene, tpm, cohort) -> gene_expression

rbind(filter(gene_expression, cohort %in% gene_disease2$cohort), 
      mutate(filter(gene_expression, cohort %in% gene_disease2$cohort), 
             cohort = 'All Cohorts')) %>% 
  inner_join(distinct(gene_disease2, cohort, Kids_First_Participant_ID, 
                      copy_number, status), 
             by = c('cohort', 'Kids_First_Participant_ID')) %>% 
  mutate(status = ifelse(!is.na(status), status,
                         case_when(copy_number == 0 ~ 'deep deletion',
                         copy_number < ploidy & copy_number != 0 ~ 'loss',
                         copy_number == ploidy ~ 'neutral',
                         copy_number <= (2*ploidy) ~ 'gain',
                         copy_number > (2*ploidy) ~ 'amplification'))) %>%
  mutate(status = factor(status, levels = c('deep deletion', 'loss', 'neutral', 
                                            'gain', 'amplification'))) %>%
ggplot(aes(x = cohort, y = tpm, fill = status)) +
  geom_boxplot() +
  viridis::scale_fill_viridis(discrete = T, option = 'viridis', drop = F) +
  labs(x = 'Cohort', y = 'Transcripts Per Million (TPM)', fill = 'Copy Number') + 
  theme_classic(base_size = 16) -> ep2b
(ep2a + (ep2b + theme(legend.position = 'none'))) + 
  plot_layout(guides = 'collect') +
  plot_annotation(title = paste0('Neuroblastoma', ": ",
                                 filter(gene_ensembl, gene_symbol == i)$gene_symbol, ' (',
                                 filter(gene_ensembl, gene_symbol == i)$ensembl, ')')) -> ep2
print(ep2)
  # ggsave(paste0('gene_plots/evidence_plot_bar_percentages_boxplots_', i, '.png'), ep2)
  # ggsave(paste0('gene_plots/evidence_plot_bar_percentages_boxplots_UPDATED_', i, '.png'), ep2)
  # print(paste0('gene_plots/evidence_plot_bar_percentages_boxplots_', i, '.png'))
}
```

Percentage bar plot + gene expression TPMs

```{r, fig.width = 10, fig.height = 4}
for (i in gene_ensembl$gene_symbol) {
### stacked percentage boxplot
  cnv %>%
  filter(gene_symbol == i, cancer_group == 'Neuroblastoma') %>%
  mutate(status = ifelse(!is.na(status), status,
                         case_when(copy_number == 0 ~ 'deep deletion',
                         copy_number < ploidy & copy_number != 0 ~ 'loss',
                         copy_number == ploidy ~ 'neutral',
                         copy_number <= (2*ploidy) ~ 'gain',
                         copy_number > (2*ploidy) ~ 'amplification'))) %>%
  add_count(cancer_group) %>% 
  filter(n >= 3) %>%
  select(Kids_First_Participant_ID, biospecimen_id, status, copy_number, ploidy, ensembl, gene_symbol, 
         cohort) -> gene_disease

rbind(gene_disease, mutate(gene_disease, cohort = 'All Cohorts')) -> gene_disease2
  
gene_disease2 %>%
  add_count(cohort, name = 'sample_count') %>%
  count(cohort, status, sample_count) %>%
  mutate(percent = round((n / sample_count) * 100),
         label = paste0(percent, '%'),
         label = ifelse(label == '0%', '<1%', label),
         status = factor(status, levels = c('amplification', 'gain', 'neutral', 
                                            'loss', 'deep deletion'))) %>%
  group_by(cohort) %>%
  mutate(count_label = ifelse(row_number() == 1, paste0('n = ', sample_count), '')) %>%
  ungroup() %>%
ggplot(aes(x = cohort, y = percent, fill = status)) +
  geom_col() +
  viridis::scale_fill_viridis(discrete = T, option = 'viridis', direction = -1, drop = F) +
  geom_text(aes(label = label, color = status), 
            position = position_stack(vjust = 0.5)) +
  scale_color_manual(values = c('gray0', 'gray20', 'gray40', 'gray60', 
                                'gray80', 'gray100'),
                     guide = 'none', drop = F) +
  geom_text(aes(label = count_label), y = 105) +
  coord_cartesian(ylim = c(0, 107)) +
  scale_y_continuous(breaks = c(0, 20, 40, 60, 80, 100)) +
  labs(x = 'Cohort', y = 'Percent Copy Number', fill = 'Copy Number') + 
  theme_classic(base_size = 16) -> ep2a
  # print(ep2a)
  
### gene expression boxplots
expression %>%
  filter(gene == i) %>%
  add_count(cohort, cancer_group, cancer_group == 'Neuroblastoma') %>% 
  filter(n >= 3) %>%
  select(Kids_First_Participant_ID, biospecimen_id, gene, tpm, cohort) -> gene_expression

rbind(filter(gene_expression, cohort %in% gene_disease2$cohort), 
      mutate(filter(gene_expression, cohort %in% gene_disease2$cohort), 
             cohort = 'All Cohorts')) %>% 
  inner_join(distinct(gene_disease2, cohort, Kids_First_Participant_ID, 
                      copy_number, status), 
             by = c('cohort', 'Kids_First_Participant_ID')) %>% 
  mutate(status = ifelse(!is.na(status), status,
                         case_when(copy_number == 0 ~ 'deep deletion',
                         copy_number < ploidy & copy_number != 0 ~ 'loss',
                         copy_number == ploidy ~ 'neutral',
                         copy_number <= (2*ploidy) ~ 'gain',
                         copy_number > (2*ploidy) ~ 'amplification'))) %>%
   mutate(status = factor(status, levels = c('deep deletion', 'loss', 'neutral', 
                                            'gain', 'amplification'))) %>%
ggplot(aes(x = cohort, y = tpm, fill = status)) +
  geom_boxplot() +
  viridis::scale_fill_viridis(discrete = T, option = 'viridis', drop = F) +
  scale_y_log10() +
  labs(x = 'Cohort', y = 'Log10 Transcripts Per Million (TPM)', fill = 'Copy Number') + 
  theme_classic(base_size = 16) -> ep2b

(ep2a + (ep2b + theme(legend.position = 'none'))) + 
  plot_layout(guides = 'collect') +
  plot_annotation(title = paste0('Neuroblastoma', ": ",
                      filter(gene_ensembl, gene_symbol == i)$gene_symbol, ' (',
                      filter(gene_ensembl, gene_symbol == i)$ensembl, ')')) -> ep3
print(ep3)
  # ggsave(paste0('gene_plots/evidence_plot_bar_percentages_boxplots_log10_', i, '.png'), ep3)
  # ggsave(paste0('gene_plots/evidence_plot_bar_percentages_boxplots_log10_UPDATED_', i, '.png'), ep3)
  # print(paste0('gene_plots/evidence_plot_bar_percentages_boxplots_', i, '.png'))
}
```

<br>

### Disease Page Plot

Is the gene heatmap feasible? Yes - at 5% or more samples with alterations or even down to 2-2.5%

```{r}
### alteration should be present in at least 5% of samples; at what level would we get the number of genes down to a reasonable level for all diseases?
cnv %>%
  filter(ensembl %in% filter(pmtl_genes, !is.na(pmtl))$ensg_id) %>%
  left_join(select(histologies, Kids_First_Biospecimen_ID, cancer_group),
            by = c('biospecimen_id' = 'Kids_First_Biospecimen_ID')) %>%
  add_count(cancer_group, name = 'subject_count') %>%
  filter(copy_number != 'ploidy') %>%
  add_count(cancer_group, ensembl) %>% 
  mutate(percent = n / subject_count) %>%
  filter(percent >= 0.025) %>%
  distinct(cancer_group, ensembl) %>%
  count(cancer_group)
```

#### circlize plots

Needed more processing power for disease page plots - see .Rmd on refosco










<br><br>

